File: nijesmik/week05/아방가르드_타일링.js
function solution(n) {
  const mod = 1_000_000_007;
  const dp = [];

  dp[0] = 1;
  dp[1] = 1;
  dp[2] = dp[1] + dp[0] * 2;
  dp[3] = dp[2] + dp[1] * 2 + dp[0] * 5;
  dp[4] = dp[3] + dp[2] * 2 + dp[1] * 5 + dp[0] * 2;
  dp[5] = dp[4] + dp[3] * 2 + dp[2] * 5 + dp[1] * 2 + dp[0] * 2;
  for (let i = 6; i <= n; i++) {
    dp[i] = (dp[i - 1] + dp[i - 2] * 2 + dp[i - 3] * 6 + dp[i - 4] - dp[i - 6] + mod) % mod;
  }

  return dp[n];
}

File: damdam6/week05/아방가르드 타일링.java
class Solution {

    public int solution(int n) {
        int MOD = 1_000_000_007;
        long[] dp = new long[n+1];
        long[] four = new long[n+1];
        long[] five = new long[n+1];
        long[] six = new long[n+1];

        dp[0] = 1;

        for(int i=1; i <= n; i++){

            if(i - 1 >= 0){
                dp[i] = (dp[i] + dp[i-1])%MOD;
            }

            if(i - 2 >= 0){
                dp[i] = (dp[i] + dp[i-2]*2%MOD)%MOD;
            }

            if(i - 3 >= 0){
                dp[i] = (dp[i] + dp[i-3]*5%MOD)%MOD;
            }

            if(i - 4 >= 0){
                four[i] = (four[i-3] + (dp[i-4]*2)%MOD)%MOD;
                dp[i] = (dp[i] + four[i])%MOD;
            }

            if(i - 5 >= 0){
                five[i] = (five[i-3] + (dp[i-5]*2)%MOD)%MOD;
                dp[i] = (dp[i] + five[i])%MOD;
            }



            if(i - 6 >= 0){
                six[i] = (six[i-3] + (dp[i-6]*4)%MOD)%MOD;
                dp[i] = (dp[i] + six[i])%MOD;
            }

        }


        long answer = dp[n];
        return Integer.parseInt(dp[n]+"");
    }
}